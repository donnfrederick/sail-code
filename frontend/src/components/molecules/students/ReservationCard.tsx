import RoundImage from 'components/atoms/RoundImage'
// import SatisfactionEmoji from 'components/atoms/SatisfactionEmoji'
import * as ConversationModels from 'models/conversation'
import moment from 'moment'
import * as React from 'react'
import { Link } from 'react-router-dom'
import styled from 'styled-components'
import isHuawei from 'utils/isHuawei'
import resolvePath from 'utils/resolvePath'

type Status = 'request' | 'queued' | 'waiting' | 'empty'
type To = 'reservations' | 'conversations'

interface Props {
  myId?: number
  reservation?: ConversationModels.Conversation
  status: Status
  to: To
}

export default (props: Props) => {
  const { myId, reservation, status, to } = props

  const partner =
    myId && reservation
      ? reservation.users.filter(user => user.id !== myId)[0]
      : null

  const introduce = partner ? partner.introduce : null

  return (status === 'request' ||
    status === 'queued' ||
    status === 'waiting') &&
    reservation ? (
    <Link to={resolvePath.page('students', `${to}/detail/${reservation.id}`)}>
      <Container data-status={status}>
        <RoundImage
          src={
            partner ? partner.picture_url : resolvePath.image('common/user.png')
          }
          size={132}
          withBorder={true}
          badge={partner ? (partner.highly_reliable ? 'elder' : null) : null}
          code={partner ? partner.country_code : ''}
        />
        <RightSide>
          <ReservedDate>
            {getConversationDate(reservation.start_at, reservation.end_at)}
          </ReservedDate>
          <Name>{partner ? `${partner.name}` : null}</Name>
          <Introduce>
            {introduce ? `${introduce.slice(0, 36)}` : null}
          </Introduce>
        </RightSide>
      </Container>
    </Link>
  ) : status === 'empty' ? (
    <Container data-status={status}>
      <Empty>Empty</Empty>
    </Container>
  ) : null
}

const getConversationDate = (startDateTime: string, endDateTime: string) => {
  const start = moment.parseZone(startDateTime)
  const end = moment.parseZone(endDateTime)

  // return `${start.format('dddd hh:mma')} - ${end.format('hh:mma')}`
  return `${start.format('MMM Do hh:mma')} - ${end.format('hh:mma')}`
}

const verticalBorder = resolvePath.image('common/dashed-border-vertical@3x.png')
const horizontalBorder = resolvePath.image(
  'common/dashed-border-horizontal@3x.png'
)

const Container = styled.div`
  display: flex;
  justify-content: flex-end;
  align-items: center;
  width: 686px;
  height: 238px;
  box-sizing: border-box;
  border-radius: 16px;
  margin: 0 auto 24px;
  ${isHuawei()
    ? 'border: 1px solid rgba(5, 68, 102, 0.3)'
    : 'box-shadow: 0 0 40px 0 rgba(5, 68, 102, 0.3)'};

  a &[data-status='queued'],
  a &[data-status='waiting'],
  a &[data-status='request'] {
    background-image: linear-gradient(289deg, #2eb1ff, #138efd);
  }
  &[data-status='empty'] {
    pointer-events: none;
    justify-content: center;
    border-radius: 0;
    background-color: transparent;
    background-repeat: no-repeat;
    background-position: left center, right center, center top, center bottom;
    background-size: 4px 100%, 4px 100%, 100% 4px, 100% 4px;
    background-image: url(${verticalBorder}), url(${verticalBorder}),
      url(${horizontalBorder}), url(${horizontalBorder});
    box-shadow: none;
  }
`

const RightSide = styled.div`
  display: flex;
  flex-flow: column;
  justify-content: center;
  align-items: flex-start;
  width: 460px;
  height: 100%;
  margin-left: 44px;
`
const Introduce = styled.div`
  line-height: 1.5;
  width: 100%;
  font-size: 26px;
  letter-spacing: 0.5px;
  text-align: left;
  color: #ffffff;
`

const ReservedDate = styled.div`
  width: 100%;
  margin-bottom: 20px;
  font-size: 26px;
  letter-spacing: 0.5px;
  text-align: left;
  color: #ffffff;
`

const Name = styled.div`
  width: 100%;
  margin-bottom: 15px;
  font-size: 40px;
  font-weight: 500;
  letter-spacing: 0.5px;
  text-align: left;
  color: #ffffff;
`

// const BottomRightSide = styled.div`
//   display: flex;
//   height: 48px;
// `

// const RequestLabel = styled.div`
//   background-color: #ffffff;
//   color: #138efd;
//   font-size: 24px;
//   box-sizing: border-box;
//   height: 48px;
//   line-height: 48px;
//   font-weight: bold;
//   padding: 0 20px;
// `

const Empty = styled.div`
  margin-bottom: 24px;
  font-size: 48px;
  font-weight: 500;
  text-align: center;
  color: #cdd5da;
`
