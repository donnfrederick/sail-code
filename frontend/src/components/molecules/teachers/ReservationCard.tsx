import RoundImage from 'components/atoms/RoundImage'
import FlatButton from 'components/atoms/teachers/FlatButton'
import { localStorage as localStorageConstants } from 'constants/index'
import * as ConversationModels from 'models/conversation'
import * as CustomUrlSchemeModels from 'models/customUrlScheme'
import moment from 'moment'
import * as qs from 'qs'
import * as React from 'react'
import { postStartConversation } from 'reducers/conversation'
import styled from 'styled-components'
import { getDays } from 'utils/calendar'
import getAuthToken from 'utils/getAuthToken'
import isHuawei from 'utils/isHuawei'
import isWebView from 'utils/isWebView'
import resolvePath from 'utils/resolvePath'

type Status = 'queued' | 'no-reserved' | 'waiting' | 'requests'

interface Props {
  index?: number
  myId?: number
  reservation?: ConversationModels.Conversation
  status: Status
  deleteConversation?: () => void
}

export default (props: Props) => {
  const { myId, reservation, status, deleteConversation } = props

  const partner =
    myId && reservation
      ? reservation.users.filter(user => user.id !== myId)[0]
      : null
  const authToken = getAuthToken()

  const pictureUrl = partner
    ? partner.picture_url
    : reservation &&
      reservation.status === 'waiting' &&
      reservation.accepting_requests &&
      reservation.conversation_requests.length
      ? reservation.conversation_requests[0].user.picture_url
      : resolvePath.image('common/user@3x.png')

  return (
    <Container data-status={status}>
      <RoundImage
        src={pictureUrl}
        size={150}
        withBorder={partner ? true : false}
        badge={partner ? partner.grade : null}
        code={partner ? partner.country_code : ''}
      />
      {status === 'queued' && reservation ? (
        <RightSide>
          <ReservedDate>
            {getConversationDate(reservation.start_at, reservation.end_at)}
          </ReservedDate>
          <Name>{partner ? `${partner.name} さん` : null}</Name>
          <ButtonContainer>
            <FlatButton
              text="詳細を見る"
              isSelected={false}
              width={224}
              height={56}
              backgroundColor="#ffffff"
              color="#1e9dfd"
              fontSize={32}
              link={resolvePath.page(
                'teachers',
                `reservations/detail/${reservation.id}`
              )}
            />
            <FlatButton
              text="会話する"
              isSelected={false}
              isDisabled={isNotStarted(reservation)}
              width={224}
              height={56}
              backgroundColor="#ffffff"
              color="#1e9dfd"
              fontSize={32}
              link={
                isWebView()
                  ? ''
                  : resolvePath.page(
                      'teachers',
                      `conversations/${reservation.id}`
                    )
              }
              onClick={() => {
                try {
                  if (authToken) {
                    postStartConversation(authToken, reservation.id)
                  }
                } catch (error) {
                  // nothing to do
                }

                if (!isWebView()) {
                  return
                }

                const params: CustomUrlSchemeModels.Conversation = {
                  end_url: `/teachers/conversations/${reservation.id}/end`,
                  partner_name: partner ? partner.name : '',
                  partner_photo_url: partner ? partner.picture_url : '',
                  redirect_url: `/teachers/conversations/${
                    reservation.id
                  }/evaluate`,
                  sora_channel_id: reservation.channel_id,
                  user_type: 'Teacher',
                  ws_token:
                    localStorage.getItem(
                      localStorageConstants.WEB_SOCKET_TOKEN_KEY
                    ) || ''
                }

                window.location.href = `sailapp://call?${qs.stringify(params)}`
              }}
            />
          </ButtonContainer>
        </RightSide>
      ) : status === 'waiting' && reservation ? (
        <RightSide>
          <ReservedDate>
            {getConversationDate(reservation.start_at, reservation.end_at)}
          </ReservedDate>
          <Name>会話相手を探しています</Name>
          <FlatButton
            text="予約を取り消す"
            isSelected={false}
            width={470}
            height={56}
            backgroundColor="#ffffff"
            color="#fec865"
            fontSize={32}
            onClick={deleteConversation}
          />
        </RightSide>
      ) : status === 'no-reserved' ? (
        <RightSide>
          <NoReserved>海外の学生と交流しましょう</NoReserved>
          <FlatButton
            text="予約する"
            isSelected={false}
            width={362}
            height={82}
            backgroundColor="#ffffff"
            color="#8ebbde"
            fontSize={40}
            link={resolvePath.page('teachers', 'reservations/new')}
          />
        </RightSide>
      ) : status === 'requests' && reservation ? (
        <RightSide>
          <RequestDate>
            {getConversationDate(reservation.start_at, reservation.end_at)}
          </RequestDate>
          <FlatButton
            text={`${
              reservation.conversation_requests.length
            }人からリクエストが来ています`}
            isSelected={false}
            width={460}
            height={52}
            backgroundColor="#138efd"
            color="#fff"
            fontSize={28}
            link={resolvePath.page(
              'teachers',
              `requests/detail/${reservation.id}`
            )}
          />
        </RightSide>
      ) : null}
    </Container>
  )
}

const getConversationDate = (startDateTime: string, endDateTime: string) => {
  const start = moment.parseZone(startDateTime)
  const end = moment.parseZone(endDateTime)

  return `${start.format(
    `M/D (${getDays('ja')[start.day()]}) H:mm`
  )} - ${end.format('H:mm')}`
}

const isStarted = (conversation: ConversationModels.Conversation) => {
  return (
    moment.parseZone(conversation.start_at).diff(moment()) <= 0 &&
    moment.parseZone(conversation.end_at).diff(moment()) > 0
  )
}

const isNotStarted = (conversation: ConversationModels.Conversation) => {
  return !isStarted(conversation)
}

// const textColors = ['#1e9dfd', '#2ac1fd', '#2dd4e9']

const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 736px;
  height: 210px;
  border-radius: 16px;
  margin: 0 auto 24px;
  ${isHuawei()
    ? 'border: 1px solid rgba(5, 68, 102, 0.3)'
    : 'box-shadow: 0 0 20px 0 rgba(5, 68, 102, 0.3)'};

  &[data-status='queued'] {
    background-image: linear-gradient(286deg, #2eb1ff, #138efd);
  }
  &[data-status='waiting'] {
    background-image: linear-gradient(106deg, #fcc76d, #fec864);
  }
  &[data-status='no-reserved'] {
    background-image: linear-gradient(106deg, #9dc1f2, #9bd2f2);
  }
  &[data-status='requests'] {
    background-color: #fff;
    background-image: none;
  }
`

const RightSide = styled.div`
  display: flex;
  flex-flow: column;
  justify-content: center;
  align-items: center;
  width: 470px;
  height: 100%;
  margin-left: 46px;
`

const ReservedDate = styled.div`
  width: 100%;
  margin-bottom: 10px;
  font-size: 40px;
  font-weight: 500;
  letter-spacing: 0px;
  text-align: left;
  color: #ffffff;
`

const RequestDate = styled(ReservedDate)`
  margin-bottom: 40px;
  color: #405766;
`

const Name = styled.div`
  width: 100%;
  margin-bottom: 10px;
  font-size: 28px;
  font-weight: 500;
  letter-spacing: 0px;
  line-height: 1.3;
  text-align: left;
  color: #ffffff;
`

const NoReserved = styled.div`
  margin-bottom: 20px;
  font-size: 28px;
  font-weight: 500;
  letter-spacing: 0px;
  text-align: center;
  color: #ffffff;
`

const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-between;
  width: 100%;
`
