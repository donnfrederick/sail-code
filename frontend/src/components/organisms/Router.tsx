import * as Sentry from '@sentry/browser'
import ErrorPage from 'components/pages/Error'
import { createBrowserHistory } from 'history'
import * as React from 'react'
import { Provider, Store } from 'react-redux'
import { ConnectedRouter } from 'react-router-redux'
import { compose, lifecycle, withState } from 'recompose'
import createStore from 'store/createStore'
import resetViewport from 'utils/resetViewport'
import userAgent from 'utils/userAgent'

declare const gtag: any

export const history = createBrowserHistory()
export const store = createStore(history) as Store<any>

history.listen(() => {
  window.scrollTo(0, 0)
  gtag('config', 'UA-105776977-2', { page_path: history.location.pathname })
  userAgent()
  resetViewport()
})

interface Props {
  children: JSX.Element
  errors: Error | null
  setErrors: (state: any) => any
}

const isProd = process.env.NODE_ENV === 'production'

const sentry = compose(
  withState('errors', 'setErrors', null),
  isProd
    ? lifecycle<Props, {}, {}>({
        componentDidCatch(error, errorInfo) {
          this.props.setErrors(error)
          Sentry.withScope(scope => {
            Object.keys(errorInfo).forEach(key => {
              scope.setExtra(key, errorInfo[key])
            })
            Sentry.captureException(error)
          })
        }
      })
    : lifecycle<Props, {}, {}>({
        componentDidCatch(error) {
          this.props.setErrors(error)
        }
      })
)

export default sentry((props: Props) => {
  const { children, errors } = props

  return (
    <Provider store={store}>
      <ConnectedRouter history={history}>
        {errors ? <ErrorPage /> : children}
      </ConnectedRouter>
    </Provider>
  )
})
