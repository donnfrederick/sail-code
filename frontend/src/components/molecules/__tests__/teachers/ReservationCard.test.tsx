import ReservationCard from 'components/molecules/teachers/ReservationCard'
import Router from 'components/organisms/Router'
import * as React from 'react'
import * as renderer from 'react-test-renderer'

test('render <ReservationCard />', () => {
  const tree = renderer
    .create(
      <Router>
        <div>
          <ReservationCard status="queued" />
          <ReservationCard status="no-reserved" />
          <ReservationCard status="waiting" />
        </div>
      </Router>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
