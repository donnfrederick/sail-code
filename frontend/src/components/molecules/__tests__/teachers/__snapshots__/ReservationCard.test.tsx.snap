// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render <ReservationCard /> 1`] = `
<div>
  <div
    className="sc-EHOje eXYtbq"
    data-status="queued"
  >
    <div
      className="sc-bdVaJa bHgtVR"
    >
      <img
        className="sc-htpNat iiyWLF"
        data-border={false}
        height="150px"
        src="/assets/img/common/user@3x.png"
        srcSet=""
        width="150px"
      />
    </div>
  </div>
  <div
    className="sc-EHOje eXYtbq"
    data-status="no-reserved"
  >
    <div
      className="sc-bdVaJa bHgtVR"
    >
      <img
        className="sc-htpNat iiyWLF"
        data-border={false}
        height="150px"
        src="/assets/img/common/user@3x.png"
        srcSet=""
        width="150px"
      />
    </div>
    <div
      className="sc-bZQynM hgsYro"
    >
      <div
        className="sc-iwsKbI hFhfOD"
      >
        海外の学生と交流しましょう
      </div>
      <button
        className="sc-bxivhb fTmgeZ"
        color="#8ebbde"
        data-disabled={false}
        data-selected={false}
        fontSize={40}
        height={82}
        width={362}
      >
        <a
          className="sc-ifAKCX XJNxd"
          href="/teachers/reservations/new"
          onClick={[Function]}
        >
          予約する
        </a>
      </button>
    </div>
  </div>
  <div
    className="sc-EHOje eXYtbq"
    data-status="waiting"
  >
    <div
      className="sc-bdVaJa bHgtVR"
    >
      <img
        className="sc-htpNat iiyWLF"
        data-border={false}
        height="150px"
        src="/assets/img/common/user@3x.png"
        srcSet=""
        width="150px"
      />
    </div>
  </div>
</div>
`;
